wcfsetup/install/files/lib/system/database/table/column/AbstractDatabaseTableColumn.class.php

```diff
@@ -0,0 +1,203 @@
+			'notNull' => $this->isNotNull() ? 1 : 0,
+			'type' => $this->getType()
+		];
+		
+		if ($this instanceof IAutoIncrementDatabaseTableColumn) {
+			$data['autoIncrement'] = $this->isAutoIncremented() ? 1 : 0;
+		}
+		
+		if ($this instanceof IDecimalsDatabaseTableColumn && $this->getDecimals() !== null) {
+			$data['decimals'] = $this->getDecimals();
+		}
+		
+		if ($this instanceof IEnumDatabaseTableColumn) {
+			$values = array_map(function($value) {
+				return str_replace(["'", '\\'], ["''", '\\\\'], $value);
+			}, $this->getEnumValues());
+			
+			$data['values'] = "'" . implode("','", $values) . "'";
+		}
+		
+		if ($this instanceof ILengthDatabaseTableColumn) {
+			$data['length'] = $this->getLength();
+		}
+		
+		if ($this instanceof IUnsignedDatabaseTableColumn) {
+			$data['unsigned'] = $this->isUnsigned();
+		}
+		
+		return $data;
+	}
+	
+	/**
+	 * @inheritDoc
+	 */
+	public function getDefaultValue() {
+		return $this->defaultValue;
+	}
+	
+	/**
+	 * @inheritDoc
+	 */
+	public function getName() {
+		if ($this->name === null) {
+			throw new \BadMethodCallException("Name of the database table column has not been set yet");
+		}
+		
+		return $this->name;
+	}
+	
+	/**
+	 * @inheritDoc
+	 */
+	public function getType() {
+		if ($this->type === null) {
+			throw new \BadMethodCallException("Type of the database table column " . get_class($this) . " has not been set yet");
+		}
+		
+		return $this->type;
+	}
+	
+	/**
+	 * @inheritDoc
+	 */
+	public function isNotNull() {
+		return $this->notNull;
+	}
+	
+	/**
+	 * @inheritDoc
+	 */
+	public function name($name) {
+		if ($this->name !== null) {
+			throw new \BadMethodCallException("Name of the database table column has already been set.");
+		}
+		
+		$this->name = $name;
+		
+		return $this;
+	}
+	
+	/**
+	 * @inheritDoc
+	 */
+	public function notNull($notNull = true) {
+		$this->notNull = $notNull;
+		
+		return $this;
+	}
+	
+	/**
+	 * Checks if the given default value is valid.
+	 * 
+	 * @param	mixed		$defaultValue	validated default value
+	 * @throws	\InvalidArgumentException	if given default value is invalid
+	 */
+	protected function validateDefaultValue($defaultValue) {
+		// does nothing
+	}
+	
+	/**
+	 * @inheritDoc
+	 */
+	public static function create($name) {
+		return (new static())->name($name);
+	}
+	
+	/**
+	 * @inheritDoc
+	 */
+	public static function createFromData($name, array $data) {
+		$column = static::create($name)
+			->defaultValue($data['default'])
+			->notNull($data['notNull']);
+		
+		if ($column instanceof IAutoIncrementDatabaseTableColumn) {
+			$column->autoIncrement($data['autoIncrement'] ?: null);
+		}
+		
+		if ($column instanceof IDecimalsDatabaseTableColumn) {
+			$column->decimals($data['decimals'] ?: null);
+		}
+		
+		if ($column instanceof IEnumDatabaseTableColumn) {
+			$values = explode(',', $data['enumValues'] ?? []);
+			
+			$values = array_map(function($value) {
+				// trim one leading and one trailing `'`
```

Comment:
Is `$value` guaranteed to be quoted at this place? I'd rather `preg_replace("/^'(.*)'$", '\\1', $value)`.

Reply:
Yes, `$data` is from `DatabaseEditor::getColumns()`.